reana_client/utils.py

```diff
@@ -207,32 +213,133 @@ def _validate_serial_workflow_environment(workflow_steps):
-            if ":" in environment_image_tag:
-                logging.warning(
-                    "Image {} has invalid tag {}".format(
-                        environment_image, environment_image_tag
-                    )
-                )
-                sys.exit(1)
-            elif environment_image_tag in ENVIRONMENT_IMAGE_SUSPECTED_TAGS_VALIDATOR:
-                logging.warning(
-                    "Using '{}' tag is not recommended in {} image.".format(
-                        environment_image_tag, environment_image
-                    )
-                )
-                sys.exit(1)
-        else:
-            logging.warning(
-                "Image {} does not have an explicit tag.".format(step["environment"])
+        image = step["environment"]
+        image_name, image_tag = _validate_image_tag(image)
+        _image_exists(image_name, image_tag)
+        uid, gids = _get_image_uid_gids(image_name, image_tag)
+        _validate_uid_gids(uid, gids, kubernetes_uid=step.get("kubernetes_uid"))
+
+
+def _validate_image_tag(image):
+    """Validate if image tag is valid."""
+    image_name, image_tag = "", ""
+    has_warnings = False
+    if ":" in image:
+        environment = image.split(":", 1)
+        image_name, image_tag = environment[0], environment[-1]
+        if ":" in image_tag:
+            click.secho(
+                "==> ERROR: Environment image {} has invalid tag '{}'".format(
+                    image_name, image_tag
+                ),
+                err=True,
+                fg="red",
             )
             sys.exit(1)
+        elif image_tag in ENVIRONMENT_IMAGE_SUSPECTED_TAGS_VALIDATOR:
+            click.secho(
+                "==> WARNING: Using '{}' tag is not recommended in {} environment image.".format(
+                    image_tag, image_name
+                ),
+                fg="yellow",
+            )
+            has_warnings = True
+    else:
+        click.secho(
+            "==> WARNING: Environment image {} does not have an explicit tag.".format(
+                image
+            ),
+            fg="yellow",
+        )
+        has_warnings = True
+        image_name = image
+    if not has_warnings:
         click.echo(
             click.style(
-                "Image {} has correct format.".format(step["environment"]), fg="green",
+                "==> Environment image {} has correct format.".format(image),
+                fg="green",
+            )
+        )
+    return image_name, image_tag
+
+
+def _image_exists(image, tag):
+    """Verify if image exists."""
+    DOCKER_REGISTRY = "https://index.docker.io/v1/repositories/{image}/tags/{tag}"
+    docker_registry_url = DOCKER_REGISTRY.format(image=image, tag=tag)
+    # Remove traling slash if no tag was specified
+    if not tag:
+        docker_registry_url = docker_registry_url[:-1]
+    response = requests.get(docker_registry_url)
```

Comment:
Exceptions will be caught already by `load_reana_spec`, is it worth adding another try-except block?

Reply:
From the user point of view, It would be nice to see if API is failing to respond since it will be an external error and not on reana side but at the same time it's okay for me to just use common exception from `load_reana_spec`